cpp/src/arrow/tensor.cc

```diff
@@ -121,4 +122,61 @@ Type::type Tensor::type_id() const { return type_->id(); }
 
 bool Tensor::Equals(const Tensor& other) const { return TensorEquals(*this, other); }
 
+// ----------------------------------------------------------------------
+// NumericTensor
+
+template <typename TYPE>
+NumericTensor<TYPE>::NumericTensor(const std::shared_ptr<Buffer>& data,
+                                   const std::vector<int64_t>& shape)
+    : NumericTensor(data, shape, {}, {}) {}
+
+template <typename TYPE>
+NumericTensor<TYPE>::NumericTensor(const std::shared_ptr<Buffer>& data,
+                                   const std::vector<int64_t>& shape,
+                                   const std::vector<int64_t>& strides)
+    : NumericTensor(data, shape, strides, {}) {}
+
+template <typename TYPE>
+NumericTensor<TYPE>::NumericTensor(const std::shared_ptr<Buffer>& data,
+                                   const std::vector<int64_t>& shape,
+                                   const std::vector<int64_t>& strides,
+                                   const std::vector<std::string>& dim_names)
+    : Tensor(TypeTraits<TYPE>::type_singleton(), data, shape, strides, dim_names) {}
+
+template <typename TYPE>
+int64_t NumericTensor<TYPE>::CalculateValueOffset(
+    const std::vector<int64_t>& index) const {
+  int64_t offset = 0;
+  if (strides_.size() > 0) {
+    for (size_t i = 0; i < index.size(); ++i) {
+      offset += index[i] * strides_[i];
+    }
+  } else {
+    const auto& fw_type = checked_cast<const FixedWidthType&>(*type_);
+    const int64_t element_size = fw_type.bit_width() / 8;
+
+    for (size_t i = 0; i < index.size() - 1; ++i) {
+      offset = index[i] + offset * shape_[i];
+    }
+    offset = (index[index.size() - 1] + offset * shape_[index.size() - 1]) * element_size;
```

Comment:
I don't get why you're special-casing `index.size() - 1` here. You could have:
```cpp
    for (size_t i = 0; i < index.size(); ++i) {
      offset = index[i] + offset * shape_[i];
    }
    offset *= element_size;
```

Reply:
Thank you for catching my bad.  I think it may be a trace of implementation trials.